glean-core/python/glean/net/ping_upload_worker.py

```diff
@@ -31,81 +32,143 @@ def storage_directory(cls) -> Path:
-
-        log.debug("Processing persisted pings at {}".format(storage_dir.resolve()))
-
-        try:
-            for path in storage_dir.iterdir():
-                if path.is_file():
-                    if cls._FILE_PATTERN.match(path.name):
-                        log.debug("Processing ping: {}".format(path.name))
-                        if not cls._process_file(path):
-                            log.error(
-                                "Error processing ping file: {}".format(path.name)
-                            )
-                            success = False
-                    else:
-                        log.debug("Pattern mismatch. Deleting {}".format(path.name))
-                        path.unlink()
-        except FileNotFoundError:
-            log.debug("File not found: {}".format(storage_dir.resolve()))
-            success = False
-
-        return success
+        from .. import Glean
+
+        if Dispatcher._testing_mode:
+            cls._test_process_sync()
+            return
+
+        if Glean._configuration._allow_multiprocessing:
+            cls._do_process_async()
+        else:
+            cls._do_process_sync()
 
     @classmethod
-    def _process_file(cls, path: Path) -> bool:
-        """
-        Processes a single ping file.
-        """
+    def _do_process_async(cls) -> "multiprocessing.Process":
         from .. import Glean
 
-        processed = False
-
-        try:
-            with path.open("r", encoding="utf-8") as fd:
-                lines = iter(fd)
-                try:
-                    url_path = next(lines).strip()
-                    serialized_ping = next(lines)
-                    valid_content = True
-                except StopIteration:
-                    valid_content = False
-            # On Windows, we must close the file before deleting it
-            if not valid_content:
-                path.unlink()
-                log.error("Invalid ping content in {}".format(path.resolve()))
-                return False
-        except FileNotFoundError:
-            log.error("Could not find ping file {}".format(path.resolve()))
-            return False
-        except IOError as e:
-            log.error("IOError when reading {}: {}".format(path.resolve(), e))
-            return False
+        # Only import the multiprocessing library if it's actually needed
+        import multiprocessing
 
-        processed = Glean._configuration.ping_uploader.do_upload(
-            url_path, serialized_ping, Glean._configuration
+        p = multiprocessing.Process(
+            target=_process_worker, args=(cls.storage_directory(), Glean._configuration)
         )
+        p.start()
+        return p
 
-        if processed:
+    @classmethod
+    def _do_process_sync(cls) -> bool:
+        from .. import Glean
+
+        return _process(cls.storage_directory(), Glean._configuration)
+
+    @classmethod
+    def _test_process_sync(cls) -> bool:
```

Comment:
This feels like something that should live in a test or test fixture, rather than in the implementation. In this case I'd personally just create a small helper function inside the test file itself.

Reply:
This needs to be here because it needs to run when the global `_testing_mode` flag is set to `True`.  `_testing_mode` is required even for Glean's users writing their own unit tests involving Glean, so it has to be in code shipped in the library.